src/cluster.c

```diff
@@ -5146,6 +5147,7 @@ void clusterSetMaster(clusterNode *n) {
     serverAssert(n != myself);
     serverAssert(myself->numslots == 0);
 
+    removeAllShardChannelSubscription();
```

Comment:
clusterSetMaster(n) doesn't imply that this node has moved to a different shard.

When a failover happens, there is a new master selected among the old replicas, but all nodes stay within the same shard. It's not necessary to cancel shard channels in this case.

Reply:
Wouldn't a reconnect be better? If the client wants to stay connected on a primary for better guarantee and lesser network overhead, it makes sense to reconnect to the primary node.

Reply:
For sharded pubsub, there is no difference between master and replica IIUC. It's possible to publish to a replica using SPUBLISH, which is then propagated to the other replicas and the master.

Isn't it better to avoid a reconnect if it's unnecessary?

Reply:
The message passed over the cluster bus link has a high overhead which some would want to avoid. #12196

Reply:
The default redirection mechanism of Redis cluster would cause a client to connect on the primary. However, after failover the client might be connected to the replica. With the disconnection, we let them retry and connect on the primary.

Reply:
With sharded-pubsub it scales better that plain old pubsub, but sure, it has a high overhead. (Hopefully it is helped by cluster v2 flotilla.) But it has the same overhead in both directions, right? A message published to master needs to be propagated to the replicas over the cluster bus too.

Reply:
Yes, that's correct.

The default subscription would have been on the primary and now it would be moved to the replica (after failover).

Reply:
> Presuming publish are made on primary, subscribing on primary has better throughput.

I fail to see how you can draw this conclusion. All messages need to be propagated to all replicas even if there are no subscribers on the replicas. The total number of messages sent over network is the same in both scenarios.

If you would say "lower latency" I would agree, but "better thoughput" I can't see.

```
(scenario 1: publish to master, subscribe to replica)

publisher      master          replica         subscribers
   |             |               |                 |
   |------------>|-------------->|================>|


(scenario 2: publish to master, subscribe to master)

publisher      master          replica         subscribers
   |             |               |                 |
   |------------>|-------------->|                 |
   |             |================================>|
```

Reply:
Yeah wrong words, updated my comment. 

@zuiderkwast what are your final thoughts? The one I'm proposing is a breaking change however I feel it might be better.

Reply:
OK, good, now I understand why. :-)

I don't really mind, but also I don't understand why sharded pubsub was implemented like that (e.g. why publish to replicas is allowed). But maybe you know because you implemented it. :D

But I have a vague idea about a possible use case, purely hypothetic. :)

Publish and subscribe to replicas is similar to readonly operations to keys on replicas, i.e. it can be used to offload reads. After slot migration, clients which read from a replica will also get a MOVED-redirect to the new master. Then they can find a replica of the new master and read from that replica.

If sharded pubsub is used the same way, i.e. if a client actually prefers to subscribe to a replica, then being unsubscribed at failover doesn't help them. If there are many subscribers, I believe subscribing from replicas can actually offload the master, because the master only needs to propagate the message once per replica. Then the replica will need to send the message to all the subscribers.

Reply:
For sharded pubsub, I wanted PUBLISH/SUBSCRIBE accessible across the shard (both primary/replica) however now in hindsight it might not have been the best decision to use clusterbus for data transfer as the metadata overhead is quite high.

I think your last explanation makes sense to me. Addressed this scenario.